crates/stark-backend/src/air_builders/symbolic/dag.rs

```diff
@@ -0,0 +1,266 @@
+use std::sync::Arc;
+
+use p3_field::Field;
+use rustc_hash::FxHashMap;
+use serde::{Deserialize, Serialize};
+
+use crate::air_builders::symbolic::{
+    symbolic_expression::SymbolicExpression, symbolic_variable::SymbolicVariable,
+};
+
+/// A node in symbolic expression DAG. Basically replace `Arc`s in `SymbolicExpression` with node IDs.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(bound = "F: Field")]
+enum SymbolicExpressionNode<F> {
+    Variable(SymbolicVariable<F>),
+    IsFirstRow,
+    IsLastRow,
+    IsTransition,
+    Constant(F),
+    Add {
+        x: usize,
+        y: usize,
```

Comment:
Maybe `left_id` and `right_id` are better to emphasize that this is not adding `x` and `y` but rather the gates.

Reply:
done